erasure-coding/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+		let mut shards = vec![
+			WrappedShard::new(vec![0; shard_len + 4]);
+			self.data_shards + self.parity_shards
+		];
+
+		for (data_chunk, blank_shard) in payload.chunks(shard_len).zip(&mut shards) {
+			let blank_shard: &mut [u8] = blank_shard.as_mut();
+			let (len_slice, blank_shard) = blank_shard.split_at_mut(4);
+			let len = ::std::cmp::min(data_chunk.len(), blank_shard.len());
+
+			// prepend the length to each data shard. this will tell us how much
+			// we need to read.
+			//
+			// this is necessary because we are doing RS encoding with 16-bit words,
+			// but the payload is a byte-slice. We need to know how much data
+			// to read from each shard when reconstructing.
+			//
+			// TODO: could be done more efficiently by pushing extra bytes onto the
+			// end. https://github.com/paritytech/polkadot/issues/88
+			(len as u32).using_encoded(|s| {
+				len_slice.copy_from_slice(s)
+			});
+
+			// fill the empty shards with the corresponding piece of the payload,
+			// zero-padded to fit in the shards.
+			blank_shard[..len].copy_from_slice(&data_chunk[..len]);
+		}
+
+		shards
+	}
+
+	// make a reed-solomon instance.
+	fn make_encoder(&self) -> ReedSolomon {
+		ReedSolomon::new(self.data_shards, self.parity_shards)
+			.expect("this struct is not created with invalid shard number; qed")
+	}
+}
+
+fn code_params(n_validators: usize) -> Result<CodeParams, Error> {
+	if n_validators > MAX_VALIDATORS { return Err(Error::TooManyValidators) }
+	if n_validators == 0 { return Err(Error::EmptyValidators) }
+
+	let n_faulty = n_validators.saturating_sub(1) / 3;
+	let n_good = n_validators - n_faulty;
+
+	Ok(CodeParams {
+		data_shards: n_faulty + 1,
+		parity_shards: n_good - 1,
+	})
+}
+
+/// Obtain erasure-coded chunks, one for each validator.
+///
+/// Works only up to 256 validators, and `n_validators` must be non-zero.
+pub fn obtain_chunks(n_validators: usize, block_data: &BlockData, extrinsic: &Extrinsic)
+	-> Result<Vec<Vec<u8>>, Error>
+{
+	let params  = code_params(n_validators)?;
+	let encoded = (block_data, extrinsic).encode();
+
+	if encoded.is_empty() {
+		return Err(Error::BadPayload);
+	}
+
+	let mut shards = params.make_shards_for(&encoded[..]);
+
+	params.make_encoder().encode(&mut shards[..])
+		.expect("Payload non-empty, shard sizes are uniform, and validator numbers checked; qed");
+
+	Ok(shards.into_iter().map(|w| w.into_inner()).collect())
+}
+
+/// Reconstruct the block data from a set of chunks.
+///
+/// Provide an iterator containing chunk data and the corresponding index.
+/// The indices of the present chunks must be indicated. If too few chunks
+/// are provided, recovery is not possible.
+///
+/// Works only up to 256 validators, and `n_validators` must be non-zero.
+pub fn reconstruct<'a, I: 'a>(n_validators: usize, chunks: I)
+	-> Result<(BlockData, Extrinsic), Error>
+	where I: IntoIterator<Item=(&'a [u8], usize)>
+{
+	let params = code_params(n_validators)?;
```

Comment:
We might worry about `n_validators` being the same on both sides here.  We also cannot trust the encoded data for `n_validators` because then an adversary can manipulate it, so I believe this is fine but maybe the comment should emphasize that `n_validators` must be correct or else we create invalid slashing attacks.

Reply:
Also I think you meant more than 256 here

Reply:
Al says the number of Validators changes in order of once a year. And it is off course retrievable from state root. He also reiterate there is no harm to add this the pieces. Your call then.